File: src/main/java/typeSpecifications/querySpecification/QueryGraph.java
package typeSpecifications.querySpecification;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

// class for specifying the query in form of a graph. it consists of QueryNodes and QueryEdges.
// so far we can add nodes or edges to it and print it.
public class QueryGraph {
    private Set<QueryNode> nodes;

    public QueryGraph() {
        nodes = new HashSet<>();
    }

    public void addQueryNodes(QueryNode... n) {
        nodes.addAll(Arrays.asList(n));
    }

    public void addQueryEdge(QueryNode source, QueryNode target, String label) {

        // .add on a hashSet does not add an element that is already present! We don't need to check whether the node is already created.

        nodes.add(source);
        nodes.add(target);

        // if this edge is already present -> don't add it.

        for (QueryEdge edge : source.getEdges()) {
            if (edge.getSource() == source && edge.getTarget() == target && edge.getLabel().equalsIgnoreCase(label)) {
                return;
            }
        }
        source.getEdges().add(new QueryEdge(source, target, label));
    }

    public void printGraph() {
        for (QueryNode node : nodes) {
            for (QueryEdge edge : node.getEdges()) {
                edge.print();
            }
        }
    }

    public Set<QueryNode> getNodes() {
        return nodes;
    }


    // only use this for test purposes!
    public boolean equals(QueryGraph otherGraph) {
        if (this.nodes == null || otherGraph.nodes == null) {
            return false;
        }

        if (this.nodes.size() != otherGraph.nodes.size()) {
            return false;
        }

        return equalsOtherNodeSet(otherGraph.nodes);
    }

    private boolean equalsOtherNodeSet(Set<QueryNode> otherNodeSet) {
        int pairsFound = 0;
        for (QueryNode thisNode : this.nodes) {
            for (QueryNode otherNode : otherNodeSet) {
                if (thisNode.equals(otherNode)) {
                    pairsFound++;
                    break;
                }
            }
        }
        return pairsFound == (this.nodes.size());
    }
}

File: tin/src/main/kotlin/tin/model/query/QueryGraph.kt
package tin.model.query

import java.util.HashSet

class QueryGraph {
    var nodes: MutableSet<QueryNode> = HashSet()

    fun addQueryNodes(vararg n: QueryNode) {
        nodes.addAll(listOf(*n))
    }

    fun addQueryEdge(source: QueryNode, target: QueryNode, label: String) {

        // .add on a hashSet does not add an element that is already present! We don't need to check whether the node is already created.
        nodes.add(source)
        nodes.add(target)

        // if this edge is already present -> don't add it.
        for (edge in source.edges) {
            if ((edge.source == source) && (edge.target == target) && edge.label.equals(label, ignoreCase = true)) {
                return
            }
        }
        source.edges.add(QueryEdge(source, target, label))
    }

    fun printGraph() {
        for (node in nodes) {
            for (edge in node.edges) {
                edge.print()
            }
        }
    }


    // only use this for test purposes!
    fun equals(otherGraph: QueryGraph): Boolean {
        return if (nodes.size != otherGraph.nodes.size) {
            false
        } else equalsOtherNodeSet(otherGraph.nodes)
    }

    private fun equalsOtherNodeSet(otherNodeSet: Set<QueryNode>): Boolean {
        var pairsFound = 0
        for (thisNode in nodes) {
            for (otherNode in otherNodeSet) {
                if (thisNode.equals(otherNode)) {
                    pairsFound++
                    break
                }
            }
        }
        return pairsFound == nodes.size
    }


}